components/merkledb/src/proof_map_index/proof_builder.rs

```diff
@@ -0,0 +1,306 @@
+    left_hashes
+}
+
+/// Nodes in the contour during creation of multi-proofs.
+#[derive(Debug)]
+struct ContourNode {
+    key: ProofPath,
+    branch: BranchNode,
+    visited_left: bool,
+    visited_right: bool,
+}
+
+impl ContourNode {
+    fn new(key: ProofPath, branch: BranchNode) -> Self {
+        Self {
+            key,
+            branch,
+            visited_left: false,
+            visited_right: false,
+        }
+    }
+
+    // Adds this contour node into a proof builder.
+    fn add_to_proof<K, V>(self, mut builder: MapProof<K, V>) -> MapProof<K, V> {
+        if !self.visited_right {
+            // This works due to the following observation: If neither of the child nodes
+            // were visited when the node is being ejected from the contour,
+            // this means that it is safe to add the left and right hashes (in this order)
+            // to the proof. The observation is provable by induction.
+            if !self.visited_left {
+                builder = builder.add_proof_entry(
+                    self.branch.child_path(ChildKind::Left),
+                    self.branch.child_hash(ChildKind::Left),
+                );
+            }
+
+            builder = builder.add_proof_entry(
+                self.branch.child_path(ChildKind::Right),
+                self.branch.child_hash(ChildKind::Right),
+            );
+        }
+        builder
+    }
+}
+
+/// Builds proofs for arbitrary set of keys in a Merkelized map.
+///
+/// This is an extension trait to `MerklePatricia`; all types that implement `MerklePatricia`
+/// (with reasonable constraints on key and value types) automatically implement
+/// `BuildProof` as well.
+pub trait BuildProof<K, V> {
+    /// Creates a proof of existence / absence for a single key.
+    fn create_proof(&self, key: K) -> MapProof<K, V>;
+
+    /// Creates a proof of existence / absence for multiple keys.
+    fn create_multi_proof(&self, keys: impl IntoIterator<Item = K>) -> MapProof<K, V>;
+}
+
+impl<K, V, T> BuildProof<K, V> for T
+where
+    K: ObjectHash,
+    T: MerklePatriciaTree<K, V>,
+{
+    fn create_proof(&self, key: K) -> MapProof<K, V> {
+        let searched_path = ProofPath::new(&key);
+
+        match self.root_node() {
+            Some((root_path, Node::Branch(root_branch))) => {
+                let mut left_hashes = Vec::with_capacity(DEFAULT_PROOF_CAPACITY);
+                let mut right_hashes = Vec::with_capacity(DEFAULT_PROOF_CAPACITY);
+
+                // Currently visited branch and its key, respectively.
+                let (mut branch, mut node_path) = (root_branch, root_path);
+
+                // Do at least one loop, even if the supplied key does not match the root key.
+                // This is necessary to put both children of the root node into the proof
+                // in this case.
+                loop {
+                    // <256 by induction; `branch` is always a branch node, and `node_path`
```

Comment:
Nit: That's correct for `ProofMapIndex` but may become incorrect for some other (probably very buggy though) implementation of the `MerklePatriciaTree` trait.